### ISSUE TITLE
Compiling the XBraid MFEM Examples

### ISSUE THREAD

**Issue by jbschroder** [mfem/mfem#2098]
Hi MFEM-ers,

It's been a while since I updated my local checkout of MFEM, and when I did, I found that the XBraid examples don't compile.  For instance in the XBraid driver's folder
$$ make drive-adv-diff-DG
....
In file included from drive-adv-diff-DG.cpp:37:
./braid_mfem.hpp:32:28: error: expected class name
class BraidVector : public HypreParVector
                           ^

I tried including linalg/hypre.hpp at the top of drive-adv-diff-DG, too, where HypreParVector is declared, but that didn't work, either.

This is probably a question best for @v-dobrev as he's worked on these drivers a lot.  

Anyway...any help is much appreciated!

Best,

Jacob

**Comment by v-dobrev** [comment]
From the error message, it looks like `HypreParVector` is not defined. Did you build MFEM with MPI support?

**Comment by jbschroder** [comment]
Hi Veselin!

Sorry about that!  I thought I had "make parallel" in my build script, but I missed it somehow.  Now, I can get
braid/drivers/drive-diffussion to work
but
braid/drivers/drive-adv-diff-DG.cpp
is causing issues.

If I run
 ./drive-adv-diff-DG --num-time-steps 256 --order 1 --ode-solver 11 --diffusion 1.0

then I get
++++++++++
....
  Braid: Begin simulation, 256 time steps


MFEM Warning: HYPRE error = 256
 ... in function: virtual void FE_Evolution::ImplicitSolve(const double, const mfem::Vector &, mfem::Vector &)
 ... in file: drive-adv-diff-DG.cpp:363

  Braid: || r_0 || not available, wall time = 8.83e-01
.....
++++++++++

I think you wrote that driver Veselin -- would you mind taking a look when you have a chance?  

Thanks!

Jacob

**Comment by v-dobrev** [comment]
Besides the "MFEM Warning", does the example actually work correctly?

**Comment by jbschroder** [comment]
Good point about the warnings -- I was going to ask about those next.  The stored regression tests for drive-diffusion.cpp still pass.  They aren't exhaustive tests, but they do make sure Braid reproduces the same residual history as before, to a good number of decimal places.  So, I would say yes, I think that drive-diffusion.cpp still works

**Comment by v-dobrev** [comment]
In that case, we can probably just silence the warnings, however, it's better to figure out what is causing these and fix the issue. The warning seems to be coming from hypre judging by the message: "MFEM Warning: HYPRE error = 256".

**Comment by jbschroder** [comment]
On a deeper inspection, none of the regression tests trigger this warning,
"MFEM Warning: HYPRE error = 256"
So, I'm actually not 100% sure that it's working. But, Braid is converging, and I'm hopeful.

But you're right, we should figure out what is causing these warnings and fix the issue.  If you aren't sure what causes hypre error 256, I can try and hook Rob in.

@rfalgout  Rob: Do you know what triggers hypre to return an error value of 256?  I tried grepping for 256, but didn't come up with anything in the hypre source files.  I believe this error value is being returned after an application of the Hypre preconditioner in MFEM.

**Comment by rfalgout** [comment]
It means “method did not converge as expected”.  If HYPRE_PRINT_ERRORS is on (`configure --with-print-errors`), it should print which file/line is setting error code.

**Comment by jbschroder** [comment]
Hi All, 

Thanks for the tips!  Here's what I found.

With hypre re-compiled --with-print-errors, I found that this error flag was being set in 
  " hypre error in file "gmres.c", line 876, error code = 256 "
This line means that GMRES has reached MaxIter, and that the residual norm is still too large.

The GMRES solver has it's tolerance set to,
" SetTol(1e-8); "
which I believe sets a relative tolerance (but correct me if I'm wrong Veselin), and the MaxIter is 1,000 (kind of a lot).

Also, its for a DG advection-diffusion problem, and this only happens when the diffusion coefficient is large  That is, I get no hypre warnings if the diffusion coefficient is small, like 0.001 or even 0.  This warning becomes more frequent as the diffusion coefficient gets larger. 

This was a little surprising, as diffusion should make things "easier", but I changed the GMRES preconditioner from ParaSails to BoomerAMG, and the warning went away.  Perhaps, the elliptic DG discretization confuses ParaSails.

So...I don't think that this was an error in the code, at all.  But rather some unanticipated behavior.  

I'm note sure what to do moving forward.  I'm mostly interested in small diffusion, so I may just leave ParaSails as the default.

I'll let you all know if I have any other questions, and Thank You for the speedy replies and help today!

- Jacob
